### PR TITLE
Allow partial downloads

### PR DESCRIPTION
**Pull request by PrestonFrom**
Add functionality to allow partial downloads of URLs. I'm not sure if you want this in the main project or not, but thought I'd submit it.

  modified:   examples/CMakeLists.txt
  new file:   examples/async_simple_max_bytes.cpp
  modified:   inc/lift/RequestHandle.h
  modified:   inc/lift/RequestPool.h
  modified:   src/RequestHandle.cpp
  modified:   src/RequestPool.cpp

### CODE REVIEW

File: inc/lift/RequestHandle.h [hunk 1]
```diff
@@ -79,6 +79,17 @@ class RequestHandle
...
     ) -> bool;
 
+    /**
+     * Sets the maximum number of bytes of data to write.
+     * (The number of bytes downloaded may be greater than the set amount,
+     * but the number of bytes written for the response's data will not
+     * exceed this amount.)
+     * @param max_bytes     The maximum number of bytes to be written for this request.
+     */
+    auto SetMaxBytes(
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: SetMaxDownloadBytes?

File: inc/lift/RequestHandle.h [hunk 2]
```diff
@@ -183,6 +194,23 @@ class RequestHandle
      */
     auto GetUserData() -> void*;
 
+    /**
+     * Set to true if the maximum number of bytes has been written.
+     * @param max_bytes_written     Boolean indicating if the max number of bytes has been written.
+     */
+    auto SetCompletedWritingMaxBytes(bool max_bytes_written) -> void;
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: is this an internal function?

File: inc/lift/RequestHandle.h [hunk 3]
```diff
@@ -198,7 +226,8 @@ class RequestHandle
         RequestPool& request_pool,
         CURL* curl_handle,
         CurlPool& curl_pool,
-        OnCompleteHandler on_complete_handler = nullptr
+        OnCompleteHandler on_complete_handler = nullptr,
+        size_t max_bytes = 0
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: max_download_bytes

File: inc/lift/RequestHandle.h [hunk 4]
```diff
@@ -266,6 +300,17 @@ class RequestHandle
     friend auto requests_accept_async(
         uv_async_t* async
     ) -> void; ///< libuv will call this function when the AddRequest() function is called.
+
+    friend auto curl_xfer_info(
+        void *p
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: comment the function as well please
void* p

File: inc/lift/RequestHandle.h [hunk 5]
```diff
@@ -266,6 +300,17 @@ class RequestHandle
...
+    ) -> int;
+
+    friend auto curl_write_partial_data(
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: comments?

File: inc/lift/RequestPool.h [hunk 6]
```diff
@@ -68,6 +68,12 @@ class RequestPool
         std::chrono::milliseconds timeout
     ) -> Request;
 
+    auto Produce(
```
Comment (jbaldwin) [inline inc/lift/RequestPool.h]: does this need a full Produce() override or can it bet a default parameter or just set after the Produce() call is done?  this feels heavy

File: src/RequestHandle.cpp [hunk 7]
```diff
@@ -65,11 +80,36 @@ auto RequestHandle::init() -> void
...
+     * the partial write function. Otherwise, use the normal write callback.
+     *
+     * Older versions of cURL don't have XFERINFOFUNCTION, so PROGRESSFUNCTION
+     * must be used in those cases. For versions that do support it, it will
+     * be preferentially used when both are set.
+     * https://curl.haxx.se/libcurl/c/progressfunc.html
+     **/
+    if (m_max_bytes)
+    {
+
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: #if LIBCURL_VERSION_NUM >= 0x072000
CURLOPT_XFERINFOFUNCTION
#else
CURLOPT_PROGRESSFUNCTION
#endif

File: src/RequestHandle.cpp [hunk 8]
```diff
@@ -175,6 +215,11 @@ auto RequestHandle::SetVersion(
     }
 }
 
+auto RequestHandle::SetMaxBytes(size_t max_bytes) -> void
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: this seems broken, as it doesn't call init() and thus the CURLOPT_XFERINFOFUNCTION might not be set

File: src/RequestHandle.cpp [hunk 9]
```diff
@@ -389,6 +445,18 @@ auto RequestHandle::setCompletionStatus(
         case CURLcode::CURLE_SSL_CONNECT_ERROR:
             m_status_code = RequestStatus::CONNECT_SSL_ERROR;
             break;
+        case CURLcode::CURLE_WRITE_ERROR:
+            /**
+             * If there is a cURL write error, but the maximum number of bytes
+             * has been written, then we intentionally aborted, so let's set
+             * this to success. Otherwise, it's a legitimate error.
+             */
+            if (GetCompletedWritingMaxBytes()) {
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: { } style, it also seems like these completed writing max bytes functions are internal...

File: src/RequestHandle.cpp [hunk 10]
```diff
@@ -480,4 +553,64 @@ auto curl_write_data(
...
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(p);
+
+    /**
+     * If WroteMaxBytes is true, we can terminate this request.
+     * Returning 1 will cause the cURL request to be aborted.
+     * (Note: This will result in the CURLcode returning as an error.
+     *  This is handled in setCompletionStatus.)
+     **/
+    if (raw_request_ptr->GetCompletedWritingMaxBytes())
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: I would suggest since this is a friend function calling into a private function of the RequestHandle so it can manipulate the max download bytes fields -- they are friends intentionally to keep the API very simple for the end user and by making these public you are leaking internal RequestHandle details through the public API

File: src/RequestPool.cpp [hunk 11]
```diff
@@ -75,6 +75,48 @@ auto RequestPool::Produce(
         return Request(this, std::move(request_handle_ptr));
     }
 }
+auto RequestPool::Produce(
```
Comment (jbaldwin) [inline src/RequestPool.cpp]: OK so it seems like you need to know max_download_bytes when producing since it affects init() and what CURLOPTs get set on the curl handle.  So I would say use an ssize_t and set it to -1 by default, which means "download the entire file", then you do not need another overload on the Produce() function but just an additional defaulted parameter

File: examples/CMakeLists.txt [hunk 12]
```diff
@@ -17,6 +17,13 @@ set(ASYNC_SIMPLE_SOURCE_FILES
 add_executable(async_simple ${ASYNC_SIMPLE_SOURCE_FILES})
```
Comment (jbaldwin) [inline examples/CMakeLists.txt]: I've added a lot of single comments, please take a look

File: inc/lift/RequestHandle.h (hunk 2 above)
Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Changed to private function (same for getCompletedWritingMaxBytes)

File: inc/lift/RequestPool.h (hunk 6 above)
Comment (PrestonFrom) [reply inc/lift/RequestPool.h]: Good point -- default parameter seems like the way to go here.

File: src/RequestHandle.cpp (hunk 8 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Totally missed that! Expanded the function to reset all the related variables and set the corresponding write callback and CURLOPT_XFERINFOFUNCTION.

File: src/RequestHandle.cpp (hunk 10 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Makes sense! Changed to private functions.

File: src/RequestPool.cpp (hunk 11 above)
Comment (PrestonFrom) [reply src/RequestPool.cpp]: That makes a lot more sense!

File: examples/CMakeLists.txt (hunk 12 above)
Comment (PrestonFrom) [reply examples/CMakeLists.txt]: Thank you for the comments! I think I got everything. I wasn't sure if it was better to create a new commit or amend and I ended up creating a new commit. In retrospect, it probably should have just been an amend. If it's difficult to read, please let me know and I'll fix it.

File: inc/lift/RequestHandle.h [hunk 13]
```diff
@@ -227,7 +215,7 @@ class RequestHandle
         CURL* curl_handle,
         CurlPool& curl_pool,
         OnCompleteHandler on_complete_handler = nullptr,
-        size_t max_bytes = 0
+        ssize_t max_download_bytes = -1
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: doc blocks actually look out of date (on_complete_handler and max_download_bytes)

File: src/RequestHandle.cpp [hunk 14]
```diff
@@ -215,9 +216,28 @@ auto RequestHandle::SetVersion(
     }
 }
 
-auto RequestHandle::SetMaxBytes(size_t max_bytes) -> void
+auto RequestHandle::SetMaxDownloadBytes(ssize_t max_download_bytes) -> void
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: So the init() function is duplicating this work, why not just have init() call this function when its setting the max_download_bytes

File: src/RequestHandle.cpp [hunk 15]
```diff
@@ -587,14 +610,24 @@ auto curl_write_partial_data(
...
 
+    size_t bytes_left_to_write;
+
+    if (raw_request_ptr->m_bytes_left_to_write > -1)
+    {
+        bytes_left_to_write = static_cast<size_t>(raw_request_ptr->m_bytes_left_to_write);
+    }
+    else
+    {
+        bytes_left_to_write = 0;
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: just initialize the variable to 0 and then you can remove the else block

File: inc/lift/RequestHandle.h [hunk 16]
```diff
@@ -254,8 +242,8 @@ class RequestHandle
 
     void* m_user_data;                          ///< The user data.
 
-    size_t m_max_bytes;                         ///< Maximum number of bytes to be written.
-    size_t m_bytes_left_to_write;                ///< Difference between maximum bytes to write and amount written
+    ssize_t m_max_download_bytes;                         ///< Maximum number of bytes to be written.
+    ssize_t m_bytes_left_to_write;               ///< Difference between maximum bytes to write and amount written
     bool m_wrote_max_bytes;                     ///< True if max number of bytes have been written
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: do we really need a bool value for this? its can be calculated by subtracting two fields correct?  I _think_ just having m_max_download_bytes and m_total_bytes_received should be enough to implement this feature with maybe some helper private member functions

File: src/RequestHandle.cpp (hunk 15 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Done!

File: src/RequestHandle.cpp (hunk 14 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Good point! Added call to init() and removed duplicate code.

File: inc/lift/RequestHandle.h (hunk 16 above)
Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Good point! Cut down to only m_max_download_bytes and m_bytes_written and add getBytesLeftToWrite.

File: inc/lift/RequestHandle.h (hunk 13 above)
Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Updated!

File: examples/CMakeLists.txt [hunk 17]
```diff
@@ -17,6 +17,13 @@ set(ASYNC_SIMPLE_SOURCE_FILES
...
 target_link_libraries(async_simple PRIVATE lifthttp -Wl,-\( ${LIBSSL} ${LIBCRYPTO} ${LIBCURL} ${LIBCARES} -Wl,-\) uv pthread z dl)
 
+### async_simple_max_bytes ###
+set(ASYNC_SIMPLE_SOURCE_FILES
```
Comment (jbaldwin) [inline examples/CMakeLists.txt]: this is resetting the ASYNC_SIMPLE_SOURCE_FILES variable, shouldn't it be ASYNC_SIMPLE_MAX_BYTES_SOURCE_FILES?

File: inc/lift/RequestHandle.h [hunk 18]
```diff
@@ -79,6 +79,17 @@ class RequestHandle
...
 
+    /**
+     * Sets the maximum number of bytes of data to write.
+     * (The number of bytes downloaded may be greater than the set amount,
+     * but the number of bytes written for the response's data will not
+     * exceed this amount.)
+     * @param max_download_bytes     The maximum number of bytes to be written for this request.
+     */
+    auto SetMaxDownloadBytes(
+            ssize_t max_download_bytes
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: nitpicking, but this looks like its tabbed in twice

File: inc/lift/RequestHandle.h [hunk 19]
```diff
@@ -191,14 +202,16 @@ class RequestHandle
      * @param request_pool The request pool that generated this handle.
      * @param curl_handle  The CURL* handle for this Request.
      * @param curl_pool    The CurlPool to return the CURL* handle when this request destructs.
+     * @param max_download_bytes    The maximum number of bytes to download, if -1, will download entire file.
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: on_complete_handler? I know it isn't from your code change but it was missing from the doc blocks previously

File: src/RequestHandle.cpp [hunk 20]
```diff
@@ -175,6 +220,14 @@ auto RequestHandle::SetVersion(
     }
 }
 
+auto RequestHandle::SetMaxDownloadBytes(ssize_t max_download_bytes) -> void
+{
+    m_max_download_bytes = max_download_bytes;
+    m_bytes_written = 0;
+
+    init();
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: hmm, this is the opposite of what I suggested, this will do a lot of duplicate work by now resetting a lot of fields on the curl handle that are unrelated to the progress/xfer functions.  I would like to see the reverse as if this is getting set on thousands of handles it _is_ a performance concern

File: src/RequestHandle.cpp [hunk 21]
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
...
+    return 0;
+}
+
+auto curl_xfer_info_new(
+    void* pointer_to_request,
+    curl_off_t, curl_off_t, curl_off_t, curl_off_t
+) -> int
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(pointer_to_request);
+
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: return raw_request_ptr->hasDownloadedRequiredBytes() -> int;  This function is duplicated with its sister function

File: src/RequestHandle.cpp [hunk 22]
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_xfer_info_old(
+    void* pointer_to_request,
+    double, double , double , double
+) -> int
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(pointer_to_request);
+
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: return raw_request_ptr->hasDownloadedRequiredBytes() -> int;  This function is duplicated with its sister function

File: src/RequestHandle.cpp [hunk 23]
```diff
@@ -401,6 +470,19 @@ auto RequestHandle::onComplete() -> void {
     m_on_complete_handler(std::move(request));
 }
 
+
+auto RequestHandle::getBytesLeftToWrite() -> size_t
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: getDownloadBytesRemaining?  getDownloadBytesLeft?  .. getRemainingDownloadBytes? I know curl uses "write" but its a confusing term in my opinion, is that the number of bytes to write in the REQUEST or "read" in the RESPONSE?

File: src/RequestPool.cpp [hunk 24]
```diff
@@ -71,6 +74,7 @@ auto RequestPool::Produce(
         request_handle_ptr->SetOnCompleteHandler(on_complete_handler);
         request_handle_ptr->SetUrl(url);
         request_handle_ptr->SetTimeout(timeout);
+        request_handle_ptr->SetMaxDownloadBytes(max_download_bytes);
```
Comment (jbaldwin) [inline src/RequestPool.cpp]: here is your performance penalty

File: src/RequestHandle.cpp [hunk 25]
```diff
@@ -389,6 +443,21 @@ auto RequestHandle::setCompletionStatus(
         case CURLcode::CURLE_SSL_CONNECT_ERROR:
             m_status_code = RequestStatus::CONNECT_SSL_ERROR;
             break;
+        case CURLcode::CURLE_WRITE_ERROR:
+            /**
+             * If there is a cURL write error, but the maximum number of bytes
+             * has been written, then we intentionally aborted, so let's set
+             * this to success. Otherwise, it's a legitimate error.
+             */
+            if (getBytesLeftToWrite() == 0)
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: I would suggest this feels like a good use of a ternary since it is only assigning m_status_code

m_status_code = (getRemainingDownloadBytes() == 0)
    ? RequestStatus::SUCCESS
    : RequestStatus::ERROR

File: src/RequestHandle.cpp [hunk 26]
```diff
@@ -401,6 +470,19 @@ auto RequestHandle::onComplete() -> void {
     m_on_complete_handler(std::move(request));
 }
 
+
+auto RequestHandle::getBytesLeftToWrite() -> size_t
+{
+    if (m_max_download_bytes > -1)
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: do you want to do a bounds check and make sure m_bytes_written is not greater than m_max_download_bytes? otherwise you'll underflow (if there is a bug somewhere) and return a very large number

File: src/RequestHandle.cpp [hunk 27]
```diff
@@ -480,4 +562,83 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_xfer_info_old(
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: this isn't that important but maybe just name them after the actual curl functions like all the rest of the curl_ functions? e.g. curl_xfer_info and curl_progress.  Unless if you know which one is the old/new its a bit ambiguous

File: src/RequestHandle.cpp [hunk 28]
```diff
@@ -65,11 +84,37 @@ auto RequestHandle::init() -> void
...
+    /**
+     * If m_max_bytes is set, then we need to set the XFERINFOFUNCTION and use
+     * the partial write function. Otherwise, use the normal write callback.
+     *
+     * Older versions of cURL don't have XFERINFOFUNCTION, so PROGRESSFUNCTION
+     * must be used in those cases. For versions that do support it, XFERINFOFUNCTION
+     * will be preferentially used when both are set.
+     * https://curl.haxx.se/libcurl/c/progressfunc.html
+     **/
+    if (m_max_download_bytes > -1)
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: SetMaxDownloadBytes(max_download_bytes) // done, the max download bytes sets all the appropriate curl_easy_setopt() fields correctly and only once

**Comment by PrestonFrom** [comment]
While working on this today, I realized we can abort the call from the write callback -- we don't need to check progress. I've removed that xferinfo/progress callbacks entirely and some of the code they were using. This should cut down on complexity.

Added bounds checking for downloaded bytes vs max bytes using an optional. I'm wondering if this has the potential to cause performance problems though...

File: src/RequestPool.cpp (hunk 24 above)
Comment (PrestonFrom) [reply src/RequestPool.cpp]: Okay, it all makes sense now!

File: src/RequestHandle.cpp (hunk 21 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

File: src/RequestHandle.cpp (hunk 22 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

File: src/RequestHandle.cpp (hunk 27 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Realized today that we don't need these callbacks to abort the CURL request, so I've removed them. More in main comment.

File: src/RequestHandle.cpp (hunk 26 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Good point! Added bounds checking and it returns an optional -- if false, we have downloaded too many bytes, otherwise it should be true and have the difference.

File: src/RequestHandle.cpp (hunk 23 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Went with getRemainingDownloadBytesOptional, since this now returns an optional as part of bounds checking.

File: src/RequestHandle.cpp (hunk 25 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Updated

File: src/RequestHandle.cpp (hunk 20 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Okay, I see where I got confused. This should be correct now.

File: src/RequestHandle.cpp (hunk 28 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Got it! That makes a lot more sense now.

File: inc/lift/RequestHandle.h (hunk 19 above)
Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Comment added

File: inc/lift/RequestHandle.h (hunk 18 above)
Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Fixed!

File: examples/CMakeLists.txt (hunk 17 above)
Comment (PrestonFrom) [reply examples/CMakeLists.txt]: Yes, it should! Fixed!

File: src/RequestStatus.cpp [hunk 29]
```diff
@@ -3,15 +3,16 @@
...
+static const std::string REQUEST_STATUS_BUILDING                    = "BUILDING";
+static const std::string REQUEST_STATUS_EXECUTING                   = "EXECUTING";
+static const std::string REQUEST_STATUS_SUCCESS                     = "SUCCESS";
+static const std::string REQUEST_STATUS_CONNECT_ERROR               = "CONNECT_ERROR";
+static const std::string REQUEST_STATUS_CONNECT_DNS_ERROR           = "CONNECT_DNS_ERROR";
+static const std::string REQUEST_STATUS_CONNECT_SSL_ERROR           = "CONNECT_SSL_ERROR";
+static const std::string REQUEST_STATUS_TIMEOUT                     = "TIMEOUT";
+static const std::string REQUEST_STATUS_RESPONSE_EMPTY              = "RESPONSE_EMPTY";
+static const std::string REQUEST_STATUS_ERROR                       = "ERROR";
+static const std::string REQUEST_STATUS_DOWNLOADED_TOO_MANY_BYTES   = "DOWNLOADED_TOO_MANY_BYTES";
```
Comment (jbaldwin) [inline src/RequestStatus.cpp]: realistically wouldn't you just chop off the extra bytes? i'm a bit confused by this error

Comment (PrestonFrom) [reply src/RequestStatus.cpp]: I was mostly trying to encapsulate an error which might occur when we write more bytes than the set max bytes -- but maybe it would be better to have this indicate a general error in the write callback function?

File: src/RequestHandle.cpp [hunk 30]
```diff
@@ -401,6 +446,19 @@ auto RequestHandle::onComplete() -> void {
     m_on_complete_handler(std::move(request));
 }
 
+auto RequestHandle::getRemainingDownloadBytesOptional() -> std::optional<size_t>
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: Optional in the function name is redundant, its in the function signature

File: src/RequestHandle.cpp [hunk 31]
```diff
@@ -389,6 +424,16 @@ auto RequestHandle::setCompletionStatus(
...
             m_status_code = RequestStatus::CONNECT_SSL_ERROR;
             break;
+        case CURLcode::CURLE_WRITE_ERROR:
+            /**
+             * If there is a cURL write error, but the maximum number of bytes has been written,
+             * then we intentionally aborted, so let's set this to success.
+             * Otherwise, we have a bug and downloaded too many bytes.
+             */
+            m_status_code = (getRemainingDownloadBytesOptional())
+                            ? RequestStatus::SUCCESS
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: whitespace

File: src/RequestStatus.cpp (hunk 29 above)
Comment (jbaldwin) [reply src/RequestStatus.cpp]: It might signify an error in your code, but realistically as an end user the only error you would get is a write error when the remote server fails and doesn't send enough bytes -- I just don't really see it as being very useful since you can always 'chop' off extra bytes if you did happen to save too many

File: src/RequestHandle.cpp [hunk 32]
```diff
@@ -480,4 +538,52 @@ auto curl_write_data(
...
+    }
+    else
+    {
+        /**
+         * If we get here, we have somehow downloaded more bytes than max bytes.
+         * Therefore, set data_length to a value not expected by CURL to cause
+         * the download to abort. The status will be set to indicate the wrong
+         * number of bytes were downloaded.
+         */
+        ++data_length;
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: this comment seems like the wrong approach, if you HAVE downloaded too many just chop them off and succeed.  realistically you shouldn't have this happen if done correctly but I don't see why you would fail if you did manage to download the requested number of bytes

File: src/RequestHandle.cpp [hunk 33]
```diff
@@ -480,4 +538,52 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_write_partial_data(
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: why not just merge curl_write_data and curl_write_partial_data? i actually can see a nice elegant way where you just keep appending and then IF max download bytes is set you can update the internal member fields and chop/bail out early if you go over

File: inc/lift/RequestHandle.h [hunk 34]
```diff
@@ -79,6 +80,17 @@ class RequestHandle
         std::chrono::milliseconds timeout
     ) -> bool;
 
+    /**
+     * Sets the maximum number of bytes of data to write.
+     * (The number of bytes downloaded may be greater than the set amount,
+     * but the number of bytes written for the response's data will not
+     * exceed this amount.)
```
Comment (jbaldwin) [inline inc/lift/RequestHandle.h]: you should mention here how to set it to "infinite" or "the full file"

Comment (PrestonFrom) [reply inc/lift/RequestHandle.h]: Added! Should I do anything to make it more prominent?

File: src/RequestHandle.cpp (hunk 31 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: I wasn't sure if this is what you had in mind -- I couldn't find another example in the code.

File: src/RequestHandle.cpp (hunk 30 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Removed!

File: src/RequestHandle.cpp (hunk 33 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Good point! Removed the partial write callback function and integrated the functionality in curl_write_data -- this also meant we didn't need to reset the the callback each time.

File: src/RequestHandle.cpp (hunk 32 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: I've updated the code and comment -- I think this is closer to what you had in mind.

File: src/RequestStatus.cpp (hunk 29 above)
Comment (PrestonFrom) [reply src/RequestStatus.cpp]: Okay, I think I'm on the right page now! Thank you for the comments!

File: inc/lift/RequestStatus.h [hunk 35]
```diff
@@ -19,7 +19,8 @@ enum class RequestStatus
     TIMEOUT,            ///< The request timed out.
     RESPONSE_EMPTY,     ///< The request has an empty response (socket severed).
 
-    ERROR               ///< The request had an error and failed.
+    ERROR,               ///< The request had an error and failed.
+    CURL_DOWNLOAD_ERROR  ///< The request had an error in the CURL write callback.
```
Comment (jbaldwin) [inline inc/lift/RequestStatus.h]: this API uses curl internally but we don't expose any "curl" names through the API, so just make it DOWNLOAD_ERROR?

File: src/RequestHandle.cpp [hunk 36]
```diff
@@ -549,66 +501,41 @@ auto curl_write_data(
...
+         * If the max number of bytes left to write is less than the
+         * data_length of the current frame, just set that as the data_length
+         * when appending to the request's data string.
+         * This will also cause the download to stop, since CURL aborts
+         * when the length returned from the write callback does match the
+         * length (size * nitems) passed in.
+         **/
+        if (bytes_left_to_write > -1)
+        {
+            auto unsigned_bytes_left_to_write = static_cast<size_t>(bytes_left_to_write);
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: "u" typically denotes unsigned, e.g. uint32_t, uint64_t, etc

File: src/RequestHandle.cpp [hunk 37]
```diff
@@ -215,9 +179,10 @@ auto RequestHandle::SetVersion(
     }
 }
 
-auto RequestHandle::SetMaxBytes(size_t max_bytes) -> void
+auto RequestHandle::SetMaxDownloadBytes(ssize_t max_download_bytes) -> void
 {
-    m_max_bytes = max_bytes;
+    m_max_download_bytes = max_download_bytes;
+    m_bytes_written = 0;
```
Comment (jbaldwin) [inline src/RequestHandle.cpp]: do you reset this field in Reset()?

File: inc/lift/RequestStatus.h (hunk 35 above)
Comment (PrestonFrom) [reply inc/lift/RequestStatus.h]: Updated to use DOWNLOAD_ERROR

File: src/RequestHandle.cpp (hunk 36 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Changed to 'ubytes_left_to_write'

File: src/RequestHandle.cpp (hunk 37 above)
Comment (PrestonFrom) [reply src/RequestHandle.cpp]: Ah, missed that!
Reset() now sets m_max_download_bytes to -1 (so it defaults to downloading everything) and m_bytes_written to 0.
